### PR TITLE
feat: finish pipeline and issue webhook

### LINKED ISSUE #3023: [Feature][plugins] The `webhook` plugins
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-devlake/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

part of https://github.com/apache/incubator-devlake/issues/2853

#### The `webhook` plugin would provide the following APIs
- Define `connection` struct as `(id, name)`, with DYNAMIC GENERATED `issue_endpoint, cicd_pipeline_endpoints`
- `POST /plugins/webhook/connections` creates `webhook connection`
- `GET /plugins/webhook/connections` get all webhook connections
- `GET /plugins/webhook/connections/:connection_id` get detail of the webhook connection
- `PATCH /plugins/webhook/connections/:connection_id` update webhook connection
- `DELETE /plugins/webhook/connections/:connection_id` delete the webhook connection
- `POST /plugins/webhook/:connection_id/cicd_pipeline` creates `cicd_pipline` and `cicd_pipeline_repos` if not existed (ci/cd webhook)
- `POST /plugins/webhook/:connection_id/issues` creates `issues` (with `type=INCIDENT` normally) and `board` if not existed (incident webhook)
- `POST /plugins/webhook/:connection_id/issues/:id/close` close specified `issue`
- Add swagger document for webhook plugin

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by likyh**
# Summary

This PR is based on #3051 .

finish pipeline and issue webhook

### Does this close any open issues?
Closes #3023

#### post a cicd task
curl http://127.0.0.1:8080/plugins/webhook/1/cicd_tasks -X 'POST' -d '{"pipeline_name":"A123","name":"unit-test","result":"IN_PROGRESS","status":"IN_PROGRESS","type":"TEST","environment":"PRODUCTION","created_date":"2020-01-01T12:00:00+00:00","finished_date":"2020-01-01T12:59:59+00:00","repo_id":"devlake","branch":"main","commit_sha":"015e3d3b480e417aede5a1293bd61de9b0fd051d"}'
```{"success":true,"message":"success"}```

#### post a cicd task but pipeline done
curl http://127.0.0.1:8080/plugins/webhook/1/cicd_tasks -X 'POST' -d '{"pipeline_name":"A123","name":"unit-test","result":"IN_PROGRESS","status":"IN_PROGRESS","type":"CI","created_date":"2020-01-01T12:00:00+00:00","finished_date":"2020-01-01T12:59:59+00:00","repo_id":"devlake","branch":"main","commit_sha":"015e3d3b480e417aede5a1293bd61de9b0fd051d"}'
```{"success":false,"message":"error executing the requested resource for plugin webhook (403)\ncaused by: can not receive this task because pipeline has already been done. (403)"}```

#### finish a cicd task
curl http://127.0.0.1:8080/plugins/webhook/1/cicd_pipeline/A123/finish -X 'POST' -d ''
```{"success":true,"message":"success"}```

#### post issue
curl http://127.0.0.1:8080/plugins/webhook/1/issues -X 'POST' -d '{"board_key":"DLK","url":"","issue_key":"DLK-1234","title":"a feature from DLK","description":"","epic_key":"","type":"BUG","status":"TODO","original_status":"created","story_point":0,"resolution_date":null,"created_date":"2020-01-01T12:00:00+00:00","updated_date":null,"lead_time_minutes":0,"parent_issue_key":"DLK-1200","priority":"","original_estimate_minutes":0,"time_spent_minutes":0,"time_remaining_minutes":0,"creator_id":"user1131","creator_name":"Nick name 1","assignee_id":"user1132","assignee_name":"Nick name 2","severity":"","component":""}'

```{"success":true,"message":"success"}```

#### close issue
curl http://127.0.0.1:8080/plugins/webhook/1/issue/DLK/DLK-1234/close -X 'POST'
```{"success":true,"message":"success","causes":null}`

### CODE REVIEW

File: plugins/webhook/api/cicd_pipeline.go [hunk 1]
```diff
@@ -18,9 +18,14 @@ limitations under the License.
 package api
```
Comment (klesh) [inline plugins/webhook/api/cicd_pipeline.go:19]: According to the latest discussion, we decided `deployment` should be a `type` of `cicd_tasks`.
We have to rename this webhook to  `cicd_tasks`

File: plugins/webhook/api/cicd_pipeline.go [hunk 2]
```diff
@@ -18,41 +18,202 @@ limitations under the License.
...
-	Id           string     `validate:"required"`
-	Result       string     `validate:"oneof=SUCCESS FAILURE ABORT"`
+type WebhookTaskRequest struct {
+	// PipelineName can be filled by any string unique in one pipeline
+	PipelineName string `mapstructure:"pipeline_name" validate:"required"`
+
+	Name         string     `validate:"required"` // Name should be unique in one pipeline
+	Result       string     `validate:"oneof=SUCCESS FAILURE ABORT IN_PROGRESS"`
 	Status       string     `validate:"oneof=IN_PROGRESS DONE"`
 	Type         string     `validate:"oneof=CI CD CI/CD"`
```
Comment (klesh) [inline plugins/webhook/api/cicd_pipeline.go:42]: The `type` field should be updated to one of `TEST，LINT，BUILD，DEPLOYMENT`, and must not be empty

File: plugins/webhook/api/cicd_pipeline.go [hunk 3]
```diff
@@ -18,41 +18,202 @@ limitations under the License.
...
 	Branch    string
 	CommitSha string `mapstructure:"commit_sha"`
 }
 
-// PostCicdPipeline
+// PostCicdTask
 // @Summary create pipeline by webhook
-// @Description Create pipeline by webhook, example: {"id":"A123123","result":"one of SUCCESS/FAILURE/ABORT","status":"one of IN_PROGRESS/DONE","type":"CI/CD","created_date":"2020-01-01T12:00:00+00:00","finished_date":"2020-01-01T12:59:59+00:00","repo":"devlake","branch":"main","commit_sha":"015e3d3b480e417aede5a1293bd61de9b0fd051d"}
+// @Description Create pipeline by webhook.<br/>
+// @Description example1: {"pipeline_name":"A123","name":"unit-test","result":"IN_PROGRESS","status":"IN_PROGRESS","type":"CI","created_date":"2020-01-01T12:00:00+00:00","finished_date":"2020-01-01T12:59:59+00:00","repo_id":"devlake","branch":"main","commit_sha":"015e3d3b480e417aede5a1293bd61de9b0fd051d"}<br/>
```
Comment (klesh) [inline plugins/webhook/api/cicd_pipeline.go:54]: The `type` field should be updated to one of `TEST，LINT，BUILD，DEPLOYMENT`, and must not be empty

File: plugins/webhook/api/cicd_pipeline.go [hunk 4]
```diff
@@ -18,41 +18,202 @@ limitations under the License.
...
+
+	return &core.ApiResourceOutput{Body: nil, Status: http.StatusOK}, nil
+}
+
+func getTypeAndResultFromTasks(domainTasks []devops.CICDTask) (typeHasCi bool, typeHasCd bool, result string) {
+	typeHasCi = false
+	typeHasCd = false
+	result = `SUCCESS`
+	for _, domainTask := range domainTasks {
+		if domainTask.Type == `CI/CD` {
```
Comment (klesh) [inline plugins/webhook/api/cicd_pipeline.go:202]: Note that the cicd_tasks Type has been changed

**Comment by likyh** [comment]
use new type(); add environment for tasks (pipeline's environment will always empty).

File: plugins/webhook/api/cicd_pipeline.go (hunk 4 above)
Comment (likyh) [reply plugins/webhook/api/cicd_pipeline.go]: fixed

File: plugins/webhook/api/cicd_pipeline.go (hunk 2 above)
Comment (likyh) [reply plugins/webhook/api/cicd_pipeline.go]: fixed

File: plugins/webhook/api/cicd_pipeline.go (hunk 3 above)
Comment (likyh) [reply plugins/webhook/api/cicd_pipeline.go]: fixed, `oneof` contains not empty

**Review by klesh — APPROVED** [review]
LGTM

**Pull request merged**